evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs

```diff
@@ -322,7 +347,114 @@ fn evm_error(retdata: &str) -> Capture<(ExitReason, Vec<u8>), Infallible> {
+    } else {
+        ExpectRevertReturn::Create(Capture::Exit((
+            ExitReason::Revert(ExitRevert::Reverted),
+            None,
+            ethers::abi::encode(&[Token::String(retdata.to_owned())]),
+        )))
+    }
+}
+
 impl<'a, 'b, B: Backend, P: PrecompileSet> CheatcodeStackExecutor<'a, 'b, B, P> {
+    fn expected_revert(
+        &mut self,
+        res: ExpectRevertReturn,
+        expected_revert: Option<Vec<u8>>,
+    ) -> ExpectRevertReturn {
+        let call = res.is_call();
+        if let Some(expected_revert) = expected_revert {
+            let data = match res {
+                ExpectRevertReturn::Create(Capture::Exit((
+                    ExitReason::Revert(_e),
+                    None,
+                    revdata,
+                ))) => Some(revdata),
+                ExpectRevertReturn::Call(Capture::Exit((ExitReason::Revert(_e), revdata))) => {
+                    Some(revdata)
+                }
+                _ => return revert_return_evm_error("Expected revert did not revert", call),
+            };
+            let final_res = if let Some(data) = data {
+                if data.len() >= 4 && data[0..4] == [8, 195, 121, 160] {
+                    // its a revert string
+                    let decoded_data =
+                        ethers::abi::decode(&[ethers::abi::ParamType::Bytes], &data[4..])
+                            .expect("String error code, but not actual string");
+                    let decoded_data = decoded_data[0]
+                        .clone()
+                        .into_bytes()
+                        .expect("Can never fail because it is bytes");
+                    if decoded_data == *expected_revert {
+                        if call {
+                            return ExpectRevertReturn::Call(Capture::Exit((
+                                ExitReason::Succeed(ExitSucceed::Returned),
+                                DUMMY_OUTPUT.to_vec(),
+                            )))
+                        } else {
+                            return ExpectRevertReturn::Create(Capture::Exit((
+                                ExitReason::Succeed(ExitSucceed::Returned),
+                                Some(
+                                    Address::from_str("0000000000000000000000000000000000000001")
+                                        .unwrap(),
+                                ),
+                                Vec::new(),
+                            )))
+                        }
+                    } else {
+                        return revert_return_evm_error(
+                            &*format!(
+                                "Error != expected error: '{}' != '{}'",
+                                String::from_utf8_lossy(&decoded_data[..]),
+                                String::from_utf8_lossy(&expected_revert)
+                            ),
+                            call,
+                        )
+                    }
+                }
```

Comment:
This can also be refactored to a function to duplicate it with the `if data == *expert` code block below right? Could make the error string ("Error != expected error") be a parameter if you want to differentiate between the 2